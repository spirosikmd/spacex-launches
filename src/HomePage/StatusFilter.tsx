import React, { SyntheticEvent } from 'react';
import PropTypes from 'prop-types';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormGroup from '@material-ui/core/FormGroup';
import FormLabel from '@material-ui/core/FormLabel';
import Switch from '@material-ui/core/Switch';

interface StatusFilterProps {
  showFailed: boolean;
  showSuccessful: boolean;
  showUpcoming: boolean;
  onChange: (value: string, checked: boolean) => void;
}

const StatusFilter = ({
  showFailed,
  showSuccessful,
  showUpcoming,
  onChange,
}: StatusFilterProps) => {
  const handleChange = (event: SyntheticEvent) => {
    const target = event.currentTarget as HTMLInputElement;
    onChange(target.value, target.checked);
  };

  return (
    <FormControl component={'fieldset' as 'div'}>
      <FormLabel component={'legend' as 'label'}>Show launches</FormLabel>
      <FormGroup row>
        <FormControlLabel
          control={
            <Switch
              checked={showFailed}
              onChange={handleChange}
              value="showFailed"
            />
          }
          label="Failed"
        />
        <FormControlLabel
          control={
            <Switch
              checked={showSuccessful}
              onChange={handleChange}
              value="showSuccessful"
            />
          }
          label="Successful"
        />
        <FormControlLabel
          control={
            <Switch
              checked={showUpcoming}
              onChange={handleChange}
              value="showUpcoming"
            />
          }
          label="Upcoming"
        />
      </FormGroup>
    </FormControl>
  );
};

StatusFilter.propTypes = {
  showFailed: PropTypes.bool.isRequired,
  showSuccessful: PropTypes.bool.isRequired,
  showUpcoming: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default React.memo(StatusFilter);
