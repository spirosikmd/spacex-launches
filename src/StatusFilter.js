import React from 'react';
import PropTypes from 'prop-types';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormGroup from '@material-ui/core/FormGroup';
import FormLabel from '@material-ui/core/FormLabel';
import Switch from '@material-ui/core/Switch';

export const StatusFilter = ({
  showFailed,
  showSuccessful,
  showUpcoming,
  onChange,
}) => {
  const handleChange = event => {
    onChange(event.currentTarget.value, event.currentTarget.checked);
  };

  return (
    <FormControl component="fieldset">
      <FormLabel component="legend">Show launches</FormLabel>
      <FormGroup row>
        <FormControlLabel
          control={
            <Switch
              checked={showFailed}
              onChange={handleChange}
              value="showFailed"
            />
          }
          label="Failed"
        />
        <FormControlLabel
          control={
            <Switch
              checked={showSuccessful}
              onChange={handleChange}
              value="showSuccessful"
            />
          }
          label="Successful"
        />
        <FormControlLabel
          control={
            <Switch
              checked={showUpcoming}
              onChange={handleChange}
              value="showUpcoming"
            />
          }
          label="Upcoming"
        />
      </FormGroup>
    </FormControl>
  );
};

StatusFilter.propTypes = {
  showFailed: PropTypes.bool.isRequired,
  showSuccessful: PropTypes.bool.isRequired,
  showUpcoming: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default React.memo(StatusFilter);
