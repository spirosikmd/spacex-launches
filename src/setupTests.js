import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { createSerializer } from 'enzyme-to-json';
import { unwrap } from '@material-ui/core/test-utils';

configure({ adapter: new Adapter() });

expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));

global.shallow = (Component, props = {}) => {
  const ComponentWithoutStyles = unwrap(Component);
  return shallow(<ComponentWithoutStyles {...props} />);
};

global.mount = (Component, props = {}) => {
  const ComponentWithoutStyles = unwrap(Component);
  return mount(<ComponentWithoutStyles {...props} />);
};

global.Date = jest.fn().mockImplementation(() => {
  return {
    toUTCString: jest.fn().mockReturnValue('Fri, 11 Jan 2019 15:31:00 GMT'),
    getDate: jest.fn().mockReturnValue(3),
    getMonth: jest.fn().mockReturnValue(10),
    getFullYear: jest.fn().mockReturnValue(2019),
  };
});

// https://github.com/airbnb/enzyme/issues/1875
jest.mock('react', () => {
  const r = jest.requireActual('react');
  return { ...r, memo: x => x };
});
