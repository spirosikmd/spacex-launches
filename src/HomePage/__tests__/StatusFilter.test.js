import React from 'react';
import { render, fireEvent } from '../../setupTests';
import StatusFilter from '../StatusFilter';

describe('StatusFilter', () => {
  let props;

  beforeEach(() => {
    props = {
      showFailed: true,
      showSuccessful: true,
      showUpcoming: true,
      onChange: jest.fn(),
    };
  });

  it('renders status filters', () => {
    const { asFragment } = render(<StatusFilter {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('when failed status changes', () => {
    it('calls the onChange prop', () => {
      const { getByLabelText } = render(<StatusFilter {...props} />);
      fireEvent.click(getByLabelText('Failed'));
      expect(props.onChange).toBeCalledWith('showFailed', false);
    });
  });

  describe('when successful status changes', () => {
    it('calls the onChange prop', () => {
      const { getByLabelText } = render(<StatusFilter {...props} />);
      fireEvent.click(getByLabelText('Successful'));
      expect(props.onChange).toBeCalledWith('showSuccessful', false);
    });
  });

  describe('when upcoming status changes', () => {
    it('calls the onChange prop', () => {
      const { getByLabelText } = render(<StatusFilter {...props} />);
      fireEvent.click(getByLabelText('Upcoming'));
      expect(props.onChange).toBeCalledWith('showUpcoming', false);
    });
  });
});
