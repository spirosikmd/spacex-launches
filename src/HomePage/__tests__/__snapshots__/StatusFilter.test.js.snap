// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusFilter renders status filters 1`] = `
<DocumentFragment>
  <fieldset
    class="MuiFormControl-root"
  >
    <legend
      class="MuiFormLabel-root"
    >
      Show launches
    </legend>
    <div
      class="MuiFormGroup-root MuiFormGroup-row"
    >
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          class="MuiSwitch-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-32 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked-33 Mui-checked"
          >
            <span
              class="MuiIconButton-label"
            >
              <span
                class="MuiSwitch-thumb"
              />
              <input
                checked=""
                class="PrivateSwitchBase-input-35 MuiSwitch-input"
                type="checkbox"
                value="showFailed"
              />
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiSwitch-track"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Failed
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          class="MuiSwitch-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-32 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked-33 Mui-checked"
          >
            <span
              class="MuiIconButton-label"
            >
              <span
                class="MuiSwitch-thumb"
              />
              <input
                checked=""
                class="PrivateSwitchBase-input-35 MuiSwitch-input"
                type="checkbox"
                value="showSuccessful"
              />
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiSwitch-track"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Successful
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          class="MuiSwitch-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-32 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked-33 Mui-checked"
          >
            <span
              class="MuiIconButton-label"
            >
              <span
                class="MuiSwitch-thumb"
              />
              <input
                checked=""
                class="PrivateSwitchBase-input-35 MuiSwitch-input"
                type="checkbox"
                value="showUpcoming"
              />
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiSwitch-track"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Upcoming
        </span>
      </label>
    </div>
  </fieldset>
</DocumentFragment>
`;
